### PR TITLE
#202 bot notification board

### PR DESCRIPTION
**Pull request by Latemus**
Added new Message Board -functionality
- Added command `/ilmoitustaulu [off]` which creates new message board or repins currently used message board message
- Message Board is concept where on bots message in a chat is used as a "host message" which content can be edited indefinately. The host message is kept pinned in the chat so that the preview of the message content is shown in the pinned message preview banner. This can be used to relay short information to users and it works as a link to the host message which can then contain more detailed information. This commit contains base implementation that can be later refined and extended.
- Contains implementation for adding 3 types of content to the board:
  - Scheduled messages are predefined scheduling for content. In this implementation the schedules are hard-coded. Scheduled message content can be chat specific or shared between chats. This commit contains an example of a possible weekly schedule.
  - Event messages are created by an predefined event in the chat. The trigger can be defined to be anything. This commit contains example of how the Twitch command starts stream status event in the chat if the chat is using message board
  - Notifications are messaegs that are shown in the message board for predefined period of time (by default 5 seconds). In this commit there are not yet implementation that uses this, as notifications by auto deleted messages is an established functionality.
- Updated dependencies (major updates listed):
  - Pillow 9.4.0 -> 10.0.1
  - PTB 20.5 -> 21.4
  - New dependency: beautifulsoup4 4.12.3 (parsing html)
- Contains major changes and/or improvements:
  - Recipe command (`/ruoka`) now randomizes random recipe from the list and fetches html content of that recipes page. Then parses the html to virtual DOM from which enriched recipe information is added to the message. New info contains: description, difficulty, prep time and serving count. This is first implementation where information is scraped from a website and can be used as an example for further implementations.
  - Major overhaul done to the Twitch Command implementation and updage loop logic. Previous implementation used recursive calls and complex hard to follow call chains. Newer logic should be more concise and clear. In addition, no recursive chains of calls are made anymore, which is more robust. In addition, support for event messages that follow the stream status added. These are used if the chat is using a message board.
  - Overhauled implementation while adding support for scheduled messages. Added sunrise and set information which is shown for only the scheduled message contents (this could be changed so that they are always shown)
    - Scheduled weather message: this is chat specific. List of users and their latest weather command city are fetched from the db. Then weather for each city is shown and rotated in the scheduled message with internal clock that updates the content every minute. The weather data is cached so that new data is fetched only once a minute.
    - Extended `handle_exception_async` introduced in the original Twitch integration implementation. Previously it worked only as function decorator to async functions. Now it works both as a function decorator and as a context manager. In addition, works for both async and sync contexts with the same call. Neat! However, the implementation is now more complex due to this update.
- Contains general fixes and small improvements. Including, but not limited to:
  - Added rate limiter which should prevent exceptions in situation where bot would exceed its Telegram bot API rate limit. Instead now the rate limiter should delay sending new updates to the API untill rolling update rate limit window allows it.
  - Updated all tests to use test_mocks_v2 module mocks. Removed test_mocks_v1 entirely, as it is was no longer used an offered inferior developer experience coparing to v2.
  - Extracted video frame capture to its own module and `VideoConvertService`. This is only minor improvement. In the previous implementation, whenever the event loop required new twitch stream frame it started a new ffmpeg subroutine and was left wating for the subprocess finishing. This could cause a problem, where multiple converts were started in the same time (multiple different streams updating at the same time). Now each convert is done sequentially from a queue with a first-in-first-out principle. Now multiple streams should not cause problems with the stream frame extracting.

### CODE REVIEW

**Comment by taturl** [comment]
Could split PR of this size to smaller PRs.

**Comment by taturl** [comment]
Apple silicon mac gets stuck when running
```
python bobweb/web/manage.py test bobweb/bob
```
On one of the ilmoitustaulu tests.

File: bobweb/bob/error_handler.py [hunk 1]
```diff
@@ -7,22 +7,45 @@
...
 
-async def error_handler(update: object, context: ContextTypes.DEFAULT_TYPE) -> None:
+async def unhandled_bot_exception_handler(update: object, context: ContextTypes.DEFAULT_TYPE) -> None:
     """
     General error handler for all unexpected errors. Logs error, notifies user by replying to the message
     that caught the error and sends traceback to the developer chat.
+
+    TODO: It might be best to first ask the user if contents of the error message can be shared to the developers either
+    TODO: anonymously or with the users details. And only after that would the error be sent to the developers chat with
+    TODO: appropriate level of details.
```
Comment (taturl) [inline bobweb/bob/error_handler.py:24]: This would make sense. Also, not good practice to leave todo's like this. Maybe raise an issue instead.

File: bobweb/bob/test_openai_api_utils.py [hunk 2]
```diff
@@ -366,18 +366,19 @@ class TikTokenTests(TestCase):
...
-        'cl100k_base' that is same for encoder used by Gpt models 3.5-turbo and 4.
-        """
-        text = ('Mary had a little lamb, Its fleece was white as snow (or black as coal). '
-                'And everywhere that Mary went, The lamb was sure to go. He followed her to '
-                'school one day, That was against the rule. It made the children laugh and '
-                'play To see a lamb at school.')
-        encoding = tiktoken.get_encoding(tiktoken_default_encoding_name)
-        self.assertEqual(251, len(text))
-        self.assertEqual(59, len(encoding.encode(text)))
+    # TODO: Tiktoken disabled. WIll be removed later
```
Comment (taturl) [inline bobweb/bob/test_openai_api_utils.py:369]: Could remove already I guess.

**Review by taturl — APPROVED** [review]

File: bobweb/bob/error_handler.py (hunk 1 above)
Comment (Latemus) [reply bobweb/bob/error_handler.py:24]: Tästä tehty issue https://github.com/M4R774/bobweb2/issues/277

**Comment by Latemus** [comment]
Ilmoitustaulu-testi korjattu. Jätän ton TikToken kirjaston siivoamisen seuraavaan merge-requestiin ja jätän nyt splittaamatta tätä. Seuraava on sit malillisempi.
